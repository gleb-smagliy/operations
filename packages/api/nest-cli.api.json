{
  "collection": "@nestjs/schematics",
  "sourceRoot": "src",
  "compilerOptions": {
    "plugins": ["@nestjs/swagger"]
  },
  "entryFile": "api"
}
