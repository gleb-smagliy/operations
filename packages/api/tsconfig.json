{
  "compilerOptions": {
    "module": "commonjs",
    "declaration": true,
    "removeComments": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "allowSyntheticDefaultImports": true,
    "target": "es2017",
    "sourceMap": true,
    "moduleResolution": "node",
    "outDir": "./dist",
    "baseUrl": "./",
    "incremental": true,
    "skipLibCheck": true
  }
}
