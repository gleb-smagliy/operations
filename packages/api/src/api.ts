import { NestFactory } from '@nestjs/core';
import { SwaggerModule, DocumentBuilder } from '@nestjs/swagger';
import { ApiModule } from './api.module';

const isDev = !process.env.NODE_ENV || process.env.NODE_ENV == 'development';

async function bootstrap() {
  const app = await NestFactory.create(ApiModule);
  
  if(isDev) {
    app.enableCors();
  }

  const config = new DocumentBuilder()
    .setTitle('Operations docs')
    .setDescription('The operations API description')
    .setVersion('1.0')
    .build();

  const document = SwaggerModule.createDocument(app, config);
  SwaggerModule.setup('api', app, document);

  await app.listen(process.env.PORT || 3030);
}
bootstrap();
