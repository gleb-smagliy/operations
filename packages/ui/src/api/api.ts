import { Api } from '@operations/api.client';

// todo: configure
const BASE_URL = 'http://localhost:3030';

export const api = new Api({
    baseUrl: BASE_URL,
});

// Define a service using a base URL and expected endpoints
