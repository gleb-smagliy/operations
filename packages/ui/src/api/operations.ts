import { createApi, fakeBaseQuery } from '@reduxjs/toolkit/query/react';
import { api } from './api';

export const operationsApi = createApi({
    reducerPath: 'operationsApi',
    baseQuery: fakeBaseQuery(),
    tagTypes: ['OPERATIONS'],
    endpoints: builder => ({ 
      getOperations: builder.query({
        async queryFn(...args: Parameters<typeof api.operation.getManyOperations>) {
          try {
            const { data } = await api.operation.getManyOperations(...args);

            return { data };
          }
          catch(err) {
            console.error(err);
            return {
                error: 'Something went wrong'
            };
          }
        },
        providesTags: ['OPERATIONS']
      }),
      createOperation: builder.mutation({
        async queryFn(...args: Parameters<typeof api.operation.createOneOperation>) {
            try {
                const { data } = await api.operation.createOneOperation(...args);

                return { data };
            }
            catch(err) {
                console.error(err);
                return {
                    error: 'Something went wrong'
                };
            }
        },
        invalidatesTags: ['OPERATIONS']
      })
    }),
  })
  
// Export hooks for usage in functional components, which are
// auto-generated based on the defined endpoints
export const {
    useGetOperationsQuery,
    useCreateOperationMutation
} = operationsApi;