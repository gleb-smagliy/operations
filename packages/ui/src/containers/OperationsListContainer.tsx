import { useCallback, useEffect, useState } from 'react';
import { useGetOperationsQuery } from '../services/operations';
import { CreateOperationModalContainer } from './CreateOperationModalContainer';
import { OperationsList } from '../components/OperationsList';
import {  EmptyList } from '../components/EmptyList';
import { Spinner } from '../components/Spinner';
import { Error } from '../components/Error';

export function useOperationsPolling(perPage: number, page: number) {
    const [shouldPoll, setShouldPoll] = useState(false);

    const params = {
        limit: perPage,
        page: page + 1
    };
    const options = { pollingInterval: shouldPoll ? 5000 : undefined };

    const { data: pageResult, isLoading, error } = useGetOperationsQuery(params, options);
    const operations = (pageResult?.data || []);

    useEffect(() => {
        setShouldPoll(operations.some(op => op.status == 'InProgress'));

        return () => setShouldPoll(false);
    },
    [operations]);

    return { pageResult, operations, isLoading, error };
}

const PER_PAGE = 5;

export function OperationsListContainer() {
    const [page, setPage] = useState(0);
    const { pageResult, operations, isLoading, error } = useOperationsPolling(PER_PAGE, page);
    const setFirstPage = () => setPage(0);

    if(isLoading) return <Spinner />;
    if(error) return <Error error={error} />;

    const createOperationButton = (
        <CreateOperationModalContainer onCreated={() => setFirstPage()}/>
    );
    
    if(operations.length == 0 || !pageResult) {
        return <EmptyList createOperationButton={createOperationButton} />;
    }

    return <OperationsList 
        operations={operations}
        pageResult={pageResult}
        perPage={PER_PAGE}
        changePage={setPage}
        onCreated={() => setFirstPage()}
    />;
}