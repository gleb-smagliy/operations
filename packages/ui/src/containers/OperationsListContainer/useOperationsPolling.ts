import { useEffect, useState } from 'react';
import { useGetOperationsQuery } from '../../services/operations';

export function useOperationsPolling(perPage: number, page: number) {
    const [shouldPoll, setShouldPoll] = useState(false);

    const params = {
        limit: perPage,
        page: page + 1
    };
    const options = { pollingInterval: shouldPoll ? 5000 : undefined };

    const { data: pageResult, isLoading, error } = useGetOperationsQuery(params, options);

    useEffect(() => {
        setShouldPoll((pageResult?.data || []).some(op => op.status === 'InProgress'));

        return () => setShouldPoll(false);
    },
    [pageResult]);

    return { pageResult, operations: (pageResult?.data || []), isLoading, error };
}