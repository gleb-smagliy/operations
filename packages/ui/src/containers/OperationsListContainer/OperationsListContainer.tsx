import { useState, useCallback } from 'react';
import { CreateOperationModal } from '../../components/CreateOperationModal';
import { OperationsList } from '../../components/OperationsList';
import {  EmptyList } from '../../components/EmptyList';
import { Spinner } from '../../components/Spinner';
import { Error } from '../../components/Error';
import { useOperationsPolling } from './useOperationsPolling';

const PER_PAGE = 5;

export function OperationsListContainer() {
    const [page, setPage] = useState(0);
    const { pageResult, operations, isLoading, error } = useOperationsPolling(PER_PAGE, page);
    const setFirstPage = useCallback(() => setPage(0), [setPage]);

    if(isLoading) return <Spinner />;
    if(error) return <Error error={error} />;

    const createOperationButton = (
        <CreateOperationModal onCreated={setFirstPage}/>
    );
    
    if(operations.length === 0 || !pageResult) {
        return <EmptyList createOperationButton={createOperationButton} />;
    }

    return <OperationsList 
        operations={operations}
        pageResult={pageResult}
        perPage={PER_PAGE}
        changePage={setPage}
        onCreated={setFirstPage}
    />;
}