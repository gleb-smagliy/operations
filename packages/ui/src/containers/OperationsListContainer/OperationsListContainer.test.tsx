
import { ApiProvider } from '@reduxjs/toolkit/query/react';
import { fireEvent, render, screen } from '@testing-library/react';
import { OperationsListContainer } from './OperationsListContainer';
import { operationsApi } from '../../services/operations';
import { api } from '../../services/api';
import {
  baseResponse,
  pagination,
  polling
} from './OperationsListContainer.test.data';

jest.mock('../../services/api', () => ({
  api: {
    operation: {
      getManyOperations: jest.fn()
    }
  }
}));

describe('<OperationsListContainer />', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  })

  it('should render all operations on a page', async () => {
    (api.operation.getManyOperations as jest.Mock).mockResolvedValue({ data: baseResponse });

    render(
      <ApiProvider api={operationsApi}>
        <OperationsListContainer />
      </ApiProvider>
    );

    const taskInProgress = await screen.findByText(/InProgress/i);
    const taskDone = await screen.findByText(/Done/i);
    const taskFailed = await screen.findByText(/Failed/i);
    const nextPage = await screen.findByLabelText(/Next page/);

    expect(taskInProgress).toBeInTheDocument();
    expect(taskDone).toBeInTheDocument();
    expect(taskFailed).toBeInTheDocument();
    expect(nextPage).toBeDisabled();
  });

  it('should render error when something goes wrong', async () => {
    jest.spyOn(global.console, 'error').mockImplementationOnce(() => {});
    (api.operation.getManyOperations as jest.Mock).mockRejectedValue({ error: 'some error' });

    render(
      <ApiProvider api={operationsApi}>
        <OperationsListContainer />
      </ApiProvider>
    );

    const error = await screen.findByText(/Something went wrong/i);

    expect(error).toBeInTheDocument();
  });

  it('should render loading when waiting for a response', () => {
    (api.operation.getManyOperations as jest.Mock).mockRejectedValue({ error: 'some error' });

    render(
      <ApiProvider api={operationsApi}>
        <OperationsListContainer />
      </ApiProvider>
    );

    const loading = screen.getByText(/Loading.../i);

    expect(loading).toBeInTheDocument();
  });

  it('should render a single item with a title and a status', async () => {
    (api.operation.getManyOperations as jest.Mock).mockResolvedValue({
      data: baseResponse
    });

    render(
      <ApiProvider api={operationsApi}>
        <OperationsListContainer />
      </ApiProvider>
    );

    const operation = baseResponse.data[0];

    const item = await screen.findByText(operation.name);

    expect(item.parentElement).toMatchSnapshot();
  });

  it('should open a create modal when clicked on a create button', async () => {
    (api.operation.getManyOperations as jest.Mock).mockResolvedValue({
      data: baseResponse
    });

    render(
      <ApiProvider api={operationsApi}>
        <OperationsListContainer />
      </ApiProvider>
    );

    fireEvent(
      await screen.findByText('Create'),
      new MouseEvent('click', { bubbles: true, cancelable: true })
    );

    expect(await screen.findByText(/Create operation/i)).toBeInTheDocument();
  });

  it('should render pagination button enabled when api returns more than 5 items', async () => {
    (api.operation.getManyOperations as jest.Mock).mockResolvedValue(pagination.page1);

    render(
      <ApiProvider api={operationsApi}>
        <OperationsListContainer />
      </ApiProvider>
    );

    const cells = await screen.findAllByText(/operationsResponse/);
    const nextPage = await screen.findByLabelText(/Next page/);

    expect(nextPage).not.toBeDisabled();
    expect(cells).toHaveLength(5);
  });

  it('should request the next page', async () => {
    (api.operation.getManyOperations as jest.Mock)
      .mockResolvedValueOnce(pagination.page1)
      .mockResolvedValueOnce(pagination.page2);

    render(
      <ApiProvider api={operationsApi}>
        <OperationsListContainer />
      </ApiProvider>
    );

    const page1Cells = await screen.findAllByText(/page1_/);

    fireEvent(
      await screen.findByLabelText(/Next page/),
      new MouseEvent('click', { bubbles: true, cancelable: true })
    );

    const page2Cells = await screen.findAllByText(/page2_/);

    expect(page1Cells).toHaveLength(5);
    expect(page2Cells).toHaveLength(2);
  });

  it('should poll for update when there some operation is in progress', async () => {
    jest.useFakeTimers();
    (api.operation.getManyOperations as jest.Mock)
      .mockResolvedValueOnce(polling.inProgress)
      .mockResolvedValueOnce(polling.done);

    render(
      <ApiProvider api={operationsApi}>
        <OperationsListContainer />
      </ApiProvider>
    );

    const inProgress = await screen.findByText(/InProgress/);
    jest.advanceTimersByTime(10000);
    const done = await screen.findByText(/Done/);

    expect(inProgress).toBeInTheDocument();
    expect(done).toBeInTheDocument();
  });
});
