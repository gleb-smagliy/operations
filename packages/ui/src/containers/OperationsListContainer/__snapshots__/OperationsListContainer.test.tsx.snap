// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OperationsListContainer /> should render a single item with a title and a status 1`] = `
<tr
  aria-label="operation"
  class="MuiTableRow-root"
>
  <th
    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
    role="cell"
    scope="row"
  >
    operationsResponse1
  </th>
  <td
    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
  >
    Done
  </td>
</tr>
`;
